File: ball.py
"""
Classe Ball - Representa a bola no jogo ByThePong
Implementa encapsulamento com atributos privados e métodos públicos
"""

import math
import random

class Ball:
    def __init__(self, x: int, y: int, radius: int = 10, initial_speed: int = 5, max_speed: int = 12, speed_increase_factor: float = 1.05, initial_angle_range: float = math.pi/4):
        """
        Inicializa a bola com posição e propriedades físicas
        
        Args:
            x (int): Posição X inicial
            y (int): Posição Y inicial
            radius (int): Raio da bola
            initial_speed (int): Velocidade inicial da bola
            max_speed (int): Velocidade máxima permitida
            speed_increase_factor (float): Fator de aumento de velocidade por rebote
            initial_angle_range (float): amplitude do ângulo inicial (em radianos)
        """
        self.__x = x
        self.__y = y
        # posição anterior para detecção contínua de colisão (swept)
        self.__prev_x = x
        self.__prev_y = y
        self.__radius = radius
        self.__speed = initial_speed
        self.__initial_speed = initial_speed
        self.__max_speed = max_speed
        self.__speed_increase_factor = speed_increase_factor
        self.__initial_angle_range = max(0.05, min(initial_angle_range, math.pi/2))
        self.__angle = random.uniform(-self.__initial_angle_range, self.__initial_angle_range)
        self.__dx = self.__speed * math.cos(self.__angle)
        self.__dy = self.__speed * math.sin(self.__angle)
    
    @property
    def x(self) -> int:
        """Retorna a posição X da bola"""
        return self.__x
    
    @x.setter
    def x(self, value: int):
        """Define a posição X da bola"""
        self.__x = value
    
    @property
    def y(self) -> int:
        """Retorna a posição Y da bola"""
        return self.__y
    
    @y.setter
    def y(self, value: int):
        """Define a posição Y da bola"""
        self.__y = value
    
    @property
    def radius(self) -> int:
        """Retorna o raio da bola"""
        return self.__radius
    
    @property
    def speed(self) -> float:
        """Retorna a velocidade da bola"""
        return self.__speed
    
    def move(self):
        """Move a bola baseado na velocidade atual"""
        # armazena posição anterior antes de mover
        self.__prev_x = self.__x
        self.__prev_y = self.__y
        self.__x += self.__dx
        self.__y += self.__dy
    
    def bounce_x(self):
        """Inverte a direção horizontal da bola"""
        self.__dx = -self.__dx
    
    def force_direction_right(self):
        """Força a direção horizontal para a direita (positiva)"""
        if self.__dx < 0:
            self.__dx = -self.__dx
    
    def force_direction_left(self):
        """Força a direção horizontal para a esquerda (negativa)"""
        if self.__dx > 0:
            self.__dx = -self.__dx
    
    def bounce_y(self):
        """Inverte a direção vertical da bola"""
        self.__dy = -self.__dy
    
    def bounce_paddle(self, paddle_y: int, paddle_height: int):
        """
        Calcula o ângulo de rebote baseado na posição da raquete
        
        Args:
            paddle_y (int): Posição Y da raquete
            paddle_height (int): Altura da raquete
        """
        # Calcula a posição relativa na raquete (0 a 1)
        relative_intersect_y = (self.__y - paddle_y) / paddle_height
        relative_intersect_y = max(0, min(1, relative_intersect_y))
        
        # Converte para ângulo (-45 a 45 graus)
        angle = (relative_intersect_y - 0.5) * math.pi / 2
        
        # Aumenta velocidade dentro do limite
        self.__speed = min(self.__speed * self.__speed_increase_factor, self.__max_speed)
        
        # Define direção horizontal após o rebote (inverte)
        horizontal_direction = -1 if self.__dx > 0 else 1
        
        # Recalcula componentes dx/dy mantendo o novo módulo de velocidade
        self.__dx = horizontal_direction * max(3.0, abs(self.__speed * math.cos(angle)))
        self.__dy = self.__speed * math.sin(angle)
        
    def accelerate(self, factor: float):
        """
        Acelera a bola multiplicando sua velocidade e clampando ao máximo.
        Args:
            factor (float): fator > 1.0 para acelerar levemente por frame
        """
        if factor <= 1.0:
            return
        # Escala componentes
        new_dx = self.__dx * factor
        new_dy = self.__dy * factor
        # Calcula nova velocidade e aplica teto
        new_speed = (new_dx ** 2 + new_dy ** 2) ** 0.5
        if new_speed > self.__max_speed:
            # Normaliza para max_speed
            if new_speed > 0:
                scale = self.__max_speed / new_speed
                new_dx *= scale
                new_dy *= scale
                new_speed = self.__max_speed
        self.__dx = new_dx
        self.__dy = new_dy
        self.__speed = new_speed
        
    def reset(self, x: int, y: int):
        """
        Reseta a bola para uma nova posição
        
        Args:
            x (int): Nova posição X
            y (int): Nova posição Y
        """
        self.__x = x
        self.__y = y
        self.__prev_x = x
        self.__prev_y = y
        self.__speed = self.__initial_speed
        self.__angle = random.uniform(-self.__initial_angle_range, self.__initial_angle_range)
        # Sorteia direção horizontal inicial para evitar previsibilidade
        direction = random.choice([-1, 1])
        # Garante componente horizontal mínima
        base_dx = abs(self.__speed * math.cos(self.__angle))
        self.__dx = direction * max(3.0, base_dx)
        self.__dy = self.__speed * math.sin(self.__angle)
    
    def get_rect(self):
        """Retorna as coordenadas do retângulo da bola para colisão"""
        return (self.__x - self.__radius, self.__y - self.__radius, 
                self.__radius * 2, self.__radius * 2)

    # --- propriedades apenas leitura das posições anteriores ---
    @property
    def prev_x(self) -> int:
        return self.__prev_x

    @property
    def prev_y(self) -> int:
        return self.__prev_y

File: public/js/ball.js
/**
 * ByThePong - Classe Ball JavaScript
 * Mantém encapsulamento e POO equivalente à versão Python
 */

class Ball {
    /**
     * Construtor da classe Ball
     * @param {number} x - Posição inicial X
     * @param {number} y - Posição inicial Y
     * @param {number} radius - Raio da bola
     * @param {number} initialSpeed - Velocidade inicial
     */
    constructor(x, y, radius = 10, initialSpeed = 5) {
        // Atributos privados
        this._x = x;
        this._y = y;
        this._radius = radius;
        this._speed = initialSpeed;
        this._initialSpeed = initialSpeed;
        
        // Calcula ângulo inicial aleatório
        this._angle = (Math.random() - 0.5) * Math.PI / 2; // -45° a +45°
        this._dx = this._speed * Math.cos(this._angle);
        this._dy = this._speed * Math.sin(this._angle);
        
        // Configurações de física
        this._maxSpeed = 12;
        this._speedIncrease = 1.05;
        this._minSpeed = 2;
    }
    
    /**
     * Getter para posição X
     * @returns {number} Posição X atual
     */
    get x() {
        return this._x;
    }
    
    /**
     * Setter para posição X
     * @param {number} value - Nova posição X
     */
    set x(value) {
        this._x = value;
    }
    
    /**
     * Getter para posição Y
     * @returns {number} Posição Y atual
     */
    get y() {
        return this._y;
    }
    
    /**
     * Setter para posição Y
     * @param {number} value - Nova posição Y
     */
    set y(value) {
        this._y = value;
    }
    
    /**
     * Getter para o raio
     * @returns {number} Raio da bola
     */
    get radius() {
        return this._radius;
    }
    
    /**
     * Getter para a velocidade
     * @returns {number} Velocidade atual
     */
    get speed() {
        return this._speed;
    }
    
    /**
     * Getter para direção X
     * @returns {number} Velocidade em X
     */
    get dx() {
        return this._dx;
    }
    
    /**
     * Getter para direção Y
     * @returns {number} Velocidade em Y
     */
    get dy() {
        return this._dy;
    }
    
    /**
     * Move a bola baseado na velocidade atual
     */
    move() {
        this._x += this._dx;
        this._y += this._dy;
    }
    
    /**
     * Inverte a direção horizontal (rebate em X)
     */
    bounceX() {
        this._dx = -this._dx;
    }
    
    /**
     * Inverte a direção vertical (rebate em Y)
     */
    bounceY() {
        this._dy = -this._dy;
    }
    
    /**
     * Rebate a bola na raquete com cálculo de ângulo
     * @param {number} paddleY - Posição Y da raquete
     * @param {number} paddleHeight - Altura da raquete
     */
    bouncePaddle(paddleY, paddleHeight) {
        // Calcula onde a bola bateu na raquete (de -1 a 1)
        const relativeIntersectY = (this._y - (paddleY + paddleHeight / 2)) / (paddleHeight / 2);
        
        // Calcula novo ângulo baseado na posição de impacto
        const angle = relativeIntersectY * Math.PI / 4; // Máximo de 45°
        
        // Aumenta velocidade gradualmente
        this._speed = Math.min(this._speed * this._speedIncrease, this._maxSpeed);
        
        // Inverte direção horizontal e aplica novo ângulo
        this._dx = -this._dx;
        this._dy = this._speed * Math.sin(angle);
        
        // Garante velocidade mínima
        if (Math.abs(this._dx) < this._minSpeed) {
            this._dx = this._dx > 0 ? this._minSpeed : -this._minSpeed;
        }
    }
    
    /**
     * Reseta a bola para posição inicial
     * @param {number} x - Nova posição X
     * @param {number} y - Nova posição Y
     */
    reset(x, y) {
        this._x = x;
        this._y = y;
        this._speed = this._initialSpeed;
        
        // Novo ângulo aleatório
        this._angle = (Math.random() - 0.5) * Math.PI / 2;
        this._dx = this._speed * Math.cos(this._angle);
        this._dy = this._speed * Math.sin(this._angle);
        
        // Direção horizontal aleatória
        if (Math.random() < 0.5) {
            this._dx = -this._dx;
        }
    }
    
    /**
     * Retorna retângulo de colisão da bola
     * @returns {Object} Retângulo com x, y, width, height
     */
    getRect() {
        return {
            x: this._x - this._radius,
            y: this._y - this._radius,
            width: this._radius * 2,
            height: this._radius * 2,
            centerX: this._x,
            centerY: this._y,
            radius: this._radius
        };
    }
    
    /**
     * Verifica colisão com retângulo
     * @param {Object} rect - Retângulo com x, y, width, height
     * @returns {boolean} True se há colisão
     */
    collidesWithRect(rect) {
        const ballRect = this.getRect();
        
        // Verifica se há sobreposição
        return (ballRect.x < rect.x + rect.width &&
                ballRect.x + ballRect.width > rect.x &&
                ballRect.y < rect.y + rect.height &&
                ballRect.y + ballRect.height > rect.y);
    }
    
    /**
     * Desenha a bola no canvas
     * @param {CanvasRenderingContext2D} ctx - Contexto do canvas
     * @param {string} color - Cor da bola
     */
    draw(ctx, color = '#ffffff') {
        ctx.save();
        
        // Desenha bola com brilho
        ctx.fillStyle = color;
        ctx.beginPath();
        ctx.arc(this._x, this._y, this._radius, 0, Math.PI * 2);
        ctx.fill();
        
        // Adiciona brilho
        const gradient = ctx.createRadialGradient(
            this._x - this._radius / 3, 
            this._y - this._radius / 3, 
            0,
            this._x, 
            this._y, 
            this._radius
        );
        gradient.addColorStop(0, 'rgba(255, 255, 255, 0.8)');
        gradient.addColorStop(1, 'rgba(255, 255, 255, 0)');
        
        ctx.fillStyle = gradient;
        ctx.beginPath();
        ctx.arc(this._x, this._y, this._radius, 0, Math.PI * 2);
        ctx.fill();
        
        ctx.restore();
    }
    
    /**
     * Retorna uma representação em string da bola
     * @returns {string} Informações da bola
     */
    toString() {
        return `Ball(x=${this._x.toFixed(1)}, y=${this._y.toFixed(1)}, speed=${this._speed.toFixed(1)})`;
    }
    
    /**
     * Retorna dados serializados da bola
     * @returns {Object} Dados da bola
     */
    toJSON() {
        return {
            x: this._x,
            y: this._y,
            radius: this._radius,
            speed: this._speed,
            dx: this._dx,
            dy: this._dy
        };
    }
    
    /**
     * Cria uma bola a partir de dados serializados
     * @param {Object} data - Dados da bola
     * @returns {Ball} Nova instância de Ball
     */
    static fromJSON(data) {
        const ball = new Ball(data.x, data.y, data.radius, data.speed);
        // Só aplica dx/dy se presentes; caso contrário mantém valores calculados
        if (typeof data.dx === 'number') ball._dx = data.dx;
        if (typeof data.dy === 'number') ball._dy = data.dy;
        return ball;
    }
}

// Torna a classe disponível globalmente
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Ball;
} else {
    window.Ball = Ball;
}
